tests/helpers/node_pool.go

```diff
@@ -0,0 +1,193 @@
+	require.NoError(t, err, "failed to label node %s as used", selectedNodeName)
+
+	t.Logf("Acquired node '%s' for test '%s'", selectedNodeName, t.Name())
+
+	return selectedNodeName
+}
+
+// isNodeAvailable checks if a node is currently available for use by a test.
+// A node is available if it does not have the NodeUsedByLabel, or if the label
+// indicates it was used by an expired test.
+func isNodeAvailable(
+	ctx context.Context,
+	t *testing.T,
+	client klient.Client,
+	nodeName string,
+	expiryDuration time.Duration,
+) bool {
+	node, err := GetNodeByName(ctx, client, nodeName)
+	if err != nil {
+		return false
+	}
+
+	if node.Spec.Unschedulable {
+		return false
+	}
+
+	usedBy, hasUsedBy := node.Labels[NodeUsedByLabel]
+	usedFromStr, hasUsedFrom := node.Labels[NodeUsedFromLabel]
+
+	if !hasUsedBy || !hasUsedFrom {
+		return true
+	}
+
+	var usedFromUnix int64
+
+	_, parseErr := fmt.Sscanf(usedFromStr, "%d", &usedFromUnix)
+	if parseErr != nil {
+		t.Logf(
+			"Warning: Malformed timestamp for node '%s' label '%s': %v. Treating as expired.",
+			nodeName,
+			NodeUsedFromLabel,
+			err,
+		)
+
+		return true
+	}
+
+	usedFrom := time.Unix(usedFromUnix, 0)
+	if time.Since(usedFrom) > expiryDuration {
+		t.Logf(
+			"Node '%s' previously used by '%s' has expired (used %s ago). Releasing.",
+			nodeName,
+			usedBy,
+			time.Since(usedFrom).Round(time.Second),
+		)
+
+		return true
+	}
+
+	return false
+}
+
+// LabelNodeAsUsed applies the usage labels to a node.
+func labelNodeAsUsed(ctx context.Context, client klient.Client, nodeName, testName string) error {
+	node, err := GetNodeByName(ctx, client, nodeName)
+	if err != nil {
+		return fmt.Errorf("failed to get node %s: %w", nodeName, err)
+	}
+
+	if node.Labels == nil {
+		node.Labels = make(map[string]string)
+	}
+
+	node.Labels[NodeUsedByLabel] = sanitizeTestName(testName)
+	node.Labels[NodeUsedFromLabel] = fmt.Sprintf("%d", time.Now().Unix())
+
+	return client.Resources().Update(ctx, node)
+}
+
+// sanitizeTestName converts a test name to a valid Kubernetes label value.
+// Labels must be 63 characters or less and match [a-z0-9]([-a-z0-9]*[a-z0-9])?
+func sanitizeTestName(name string) string {
+	sanitized := strings.ToLower(name)
+	sanitized = strings.ReplaceAll(sanitized, "/", "-")
+	sanitized = strings.ReplaceAll(sanitized, "_", "-")
+	sanitized = strings.ReplaceAll(sanitized, " ", "-")
+
+	sanitized = strings.Trim(sanitized, "-")
+
+	if len(sanitized) > 0 && !isAlphanumeric(sanitized[0]) {
+		sanitized = "t-" + sanitized
+	}
+
+	if len(sanitized) > 0 && !isAlphanumeric(sanitized[len(sanitized)-1]) {
+		sanitized += "-t"
+	}
+
+	if len(sanitized) > 63 {
+		sanitized = sanitized[:63]
+	}
+
+	if sanitized == "" {
+		return "default-test"
+	}
+
+	return sanitized
+}
+
+// isAlphanumeric checks if a byte is an alphanumeric character
+func isAlphanumeric(c byte) bool {
+	return (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9')
```

Comment:
Can we use https://pkg.go.dev/unicode#IsLetter and https://pkg.go.dev/unicode#IsNumber instead?